README
README for xsoldier
$Id: README,v 1.9 2011/08/09 08:12:25 oohara Exp $

* what is xsoldier
xsoldier is a shoot 'em up game with the "not shooting" bonus.  It is
2D (that is, hardware acceleration is not necessary if you have a "not bad"
CPU) game for one player using the X Window System or the Simple DirectMedia
Layer (SDL) library.

So, why yet another shoot 'em up?  Because xsoldier has a unique feature
--- the "not shooting" bonus.  You can get a pretty good score by not
pressing the shoot button.  This is an answer to my motto for a shoot 'em up,
that is, "why a shoot button?".  After all, if the game is a simple
kill-everything and shooting is always a good thing, why do I have to
press a shoot button (or worse, hit the shoot button again and again)?
You can play xsoldier as a simple kill-everything (and it is the best way
to survive), but if you care your score, you need to kill all enemies
with a minimum number of bullets.

* how to install xsoldier
See the file INSTALL for information about how to install xsoldier.
In short, install libc, libX11 and libXpm, then type
1) ./configure
2) make
3) make install
If you use SDL instead of X, install libSDL and libSDL_image instead of
libX11 and libXpm, then use the --with-sdl option of ./configure .

* overview of the game
In xsoldier, you move a space ship and shoot enemies (they usually appear
from the upper side of the window, but some appear from the _lower_ side).
If you kill the boss enemy which appears at the end of each stage, or if
you can survive till the boss enemy gets bored with you and escapes, you
can go to the next stage.

xsoldier has 8 stages.  You need to clear each stage 3 times to clear the
entire game (24 stages total).  A set of the 8 stages is called a "loop".
If you cleared the first loop (that is, the first 8 stages), the second
loop will begin after the ending (a staff scroll), with more difficulty
(enemies shoot more, enemy shots move faster and so on).  Similarly,
the third loop will begin after the second loop (and the second ending).
The third loop is very dangerous because enemies will _always_ leave a shot
when they die.

* command line options

** (-d | -display) NAME
Specify the (X window) display on which you want to play the game.

** (-w | -wait) NUMBER
Set the speed of the game.  A higher number means the game runs slower.
NUMBER must be a positive integer.  If you use this option and if NUMBER
is greater than the default (use -h to see it), the ranking mode is
turned off.

** (-c | -cmap)
This option has no effect (only for backward compatibility).

** (-s | -score)
Show the top 10 soldiers, that is, the high score list.

** (-h | -help)
Show a short help text and exit.

** -maxlevel
Force max level enemy attack, making your life really harder.

** -nopausemessage
Don't show "Press [p] to resume game." message when the game is paused.
This is useful for capturing a screen snapshot.


If you compiled xsoldier with debugging features, the follwing options
are also available.  They disable the ranking mode.

** -stage L-S
Start the game with the stage S of the loop L.  S must be a positive
integer between 1 and 8, and L must be a positive integer between
1 and 3.

** -stage S
Start the game with the stage S.  S must be a positive integer between
1 and 8,

** -loop L
Start the game with the loop L.  L must be a positive integer between
1 and 3.

** -ship N
Start the game with N ships.  N must be a integer between 0 and 99.

** -maxpower
Make sure that you always have a maximum power weapon.

* key control
cursor keys        move
left shift         shoot (auto-repeat)
a                  speed up
s                  speed down
p                  pause/resume
q                  quit the game
space              start the game (in the title/ending)

You can use a joystick too if xsoldier is compiled with joystick support
(see INSTALL).  Note that joystick support is no longer maintained.
Probably it won't work on a system other than Linux.

If you compiled xsoldier with debug features, some keys are also available
for cheating:
w                  change weapon
c                  clear enemy shots

* power-up items
Some enemies sometimes leave a power-up item when they die (others never do).
There are 2 kinds of items.  The one which says "Pow" makes your shots more
powerful.  The other one (which changes its appearance periodically) changes
your weapon and makes your shots a bit more powerful.

There are 3 kinds of weapons you can use --- blue, yellow and gray.
The blue shot is fast, damages fairly well and you can shoot it
backward if you correct enough power-up items.  The yellow shot shoots
many, many bullets, so you don't need to aim very carefully.  However,
each bullet deals only a little damage.  The gray shot is slow and
you can shoot it only forward, but it damages very well (even boss enemies
will die in a few seconds).

If you already have enough power-up items and gets another one, it will
give you bonus points.  If you die, you may lose some of power-up items.

* "not shooting" bonus
You get a bonus automatically as long as you don't press the shoot key
(to be precise, if it is at least 5 frames after you released the shoot key).
This bonus is called "not shooting" bonus.  The more enemy there are in
the window, the more points you get.  Enemy shots, power-up items and
explosions also count as enemies for the "not shooting" bonus.

* shoot down bonus
When you clear a stage by destroying the boss enemy, you get a bonus
according to the percentage of enemies you destroyed in this stage.
95% destroy bonus is 1/2 of 100% destroy bonus, 90% destroy bonus is
1/3 of of 100% destroy bonus and so on.  The shoot down bonus is larger
in later stages.

* game tips
Your ship is smaller than it looks.  In other words, your ship will not be
destroyed even if enemy shots touch it a bit.

The yellow shot is the best way to survive.  Keep pressing the shoot key
unless you are an expert.  Problems are that your score will be poor and
that the yellow shot is not good at killing the boss enemies.

Play with the boss enemies for a while to get the "not shooting" bonus.
stage   estimated bonus    difficulty
  1       small              easy
  2       small              very easy (loop 1), hard (loop 2 or later)
  3       small              hard
  4       medium             normal (loop 1 or 2), hard (loop 3)
  5       very big           easy (loop 1), normal (loop 2 or later)
  6       big                hard
  7       small              hard
  8       very big           very hard
Make sure that you destroy the boss enemies before they escape.  Don't
miss the shoot down bonus.

* feedback
Send comments or bug reports to Oohara Yuuma <oohara@libra.interq.or.jp>.
Also check the xsoldier webpage:
http://www.interq.or.jp/libra/oohara/xsoldier/index.html

* license
See the file LICENSE for information about the license.  In short,
the license is GPL version 2 or later.

* history
The original author of xsoldier was Yuusuke HASHIMOTO <hachi@pm.highway.ne.jp>.
He maintained xsoldier until version 0.96.  It was a simple but decent
shoot 'em up.  The original xsoldier webpage was
http://www.surfline.ne.jp/hachi/xsoldier.html .

Oohara Yuuma <oohara@libra.interq.or.jp> took over the development of
xsoldier and added the "not shooting" bonus.  The game bacame easy if
you are interested in just surviving.


ChangeLog
2011-08-13  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 1.8
	* score.c: compare_entry() is now static
	* configure.ac: added x before with/enable variables in case
	  their values begin with a special character
	* player.c: removed the "not shooting" bonus debug message
	  because it is too noisy
	* boss.c: removed debug messages in BossHit1 because they are
	  too noisy

2011-08-10  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 1.7
	* joystick.c, joystick.h: define empty functions instead of
	  commenting them out when the joystick support is disabled
	  to avoid an empty .c file
	* enemyshot.c: included stdio.h and stdlib.h (for abs())
	* Makefile.am: removed -Wcast-qual because the file name for
	  XpmReadFileToPixmap is char *, not const char *

2011-08-09  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 1.6
	* configure.ac: fixed so that --enable options can be disabled
	  correctly
	* configure.ac: rewrite --with option
	* configure.ac: added AC_PROG_CPP because AC_CHECK_HEADERS requires it
	* configure.ac: checks the function atexit because graphic.c uses it
	* configure.ac: fixed comments
	* joystick.c, joystick.h: comment out the whole file if JSTK is
	  not defined
	* Makefile.am: fixed comment so that Makefile doesn't say "process
	  this file with automake"
	* Makefile.am, configure.ac: use the tar-ustar option of automake
	  instead of AMTAR = ustar in Makefile.am
	* configure.ac: added the foreign option to AM_INIT_AUTOMAKE
	* README: says that the joystick support is only on Linux

2009-11-08  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 1.5
	* game.c, xsoldier.h: make sure that an object never acts at the same
	  frame as it is created
	* game.c: removed unnecessary variable ocheck
	* version.h: updated UPDATE

2006-09-16  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 1.4
	* Makefile.am: minor cleanup
	  - uses some AM_variables
	  - gcc -g and the overkill warnings are now default
	  - set AMTAR to ustar
	* Makefile.in: regenerated with automake 1.9
	* configure: ./configure accepts --without-sld as well as --with-sdl
	* INSTALL: fixed high score file instruction
	* input.c, main.c, manage.c, xsolder.h: the SDL version respects
	  the TERM signal and closing the window

2002-05-26  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 1.3
	* configure.ac:  does not stop configuring even if --with-sdl is
	  specified but AC_CHECK_LIB([SDL], [SDL_Init]) fails (this won't do
	  any harm because if there is no SDL then "make" will fail correctly)
	  (I don't know why AC_CHECK_LIB fails on a non-i386 architecture
	  (perhaps because of missing -lpthread?))
	* configure.ac: regenerated
	* image.c, manage.c: s/HAVE_LIBSDL_IMAGE/HAVE_LIBSDL/g
	* common.c, ending.c, enemyshot.c, game.c, joystick.c, opening.c:
	  added some #include to avoid gcc -Wall warnings
	* enemyshot.c, graphic.c, manage.c, player.c: use division instead
	  of shift
	* score.c: fake records no longer have loop 0
	* score.c: fake records are more close to the default
	* image.c: fixed image loading so that xsoldier displays the correct
	  color on a 16-bit-per-pixel display (thanks to Stephen McCamant
	  <smcc@CSUA.Berkeley.EDU> for the patch)
	* README: fixed typo
	* image.c: SDL version sets the palette of pixmaps explicitly if
	  the screen is 8 bpp (without this the window is blank)
	* README, extern.h, graphic.c, image.c, main.c, xsoldier.6: now
	  -c option does nothing because it does not work with raw X on
	  8 bpp screen
	* image.h: SDL version uses typedef instead of #define for Image
	* graphic.c, image.c, star.c: fixed memory leak (I still don't know
	  how to free memory allocated by getpwuid() in manage.c)
	* README: fixed description (xsoldier is not first-person, "for one
	  player" is the right term)
	* boss.c: you can't go to the next stage if the game is over

2002-04-24  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 1.2
	* xsoldier is ported to SDL (the Simple DirectMedia Layer library)
	* extern.h: fixed a typo in comment
	* callback.c, game.c, graphic.c, image.c, input.c, joystick.c,
	  main.c, manage.c, config.h, extern.h, graphic.h, image.h, key.h:
	  Added libsdl support (use --with-sdl).  Note that libsdl support
	  uses a double-standard for coordinates to keep the 10 pixel
	  boundary
	* enemyshot.c: enemy shots do move
	* INSTALL, README: added note on SDL
	* README: added a game tip

2002-04-21  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 1.1
	* now xsolider comes with its own font
	* graphic.c: no longer use the X font (this is necessary for porting
	  to SDL)
	* pixmap/font*.xpm: new files, provide the font
	* boss.c, enemy.c: enemies appear more gracefully
	* game.c: fixed counter of the number of normal enemies
	* Makefile.am: "make install" does not overwrite the high score file
	* INSTALL: added more info on the high score file
	
2002-04-21  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 1.0
	* now I am the official developer of xsoldier (I got a mail from
	  the original of xsoldier, Yuusuke HASHIMOTO <hachi@pm.highway.ne.jp>)
	* player.c: weapon 1 shoots more and damages less
	* game.c: set boss time to 2000 instead of 3000 if not in final stage
	* boss.c: teleportation of boss 6 takes less time
	* enemy.c: power-up items move faster
	* boss.c: boss has a fixed (independent of loop) hit point
	* game.c, player.c, player.h: you lose some of power-up items
	  when you are killed
	* boss.c: reduced the amount of ring shots of the last boss
	* boss.c: the last boss shoots a 5way shot in loop 1
	* enemy.c: enemy 6 may give you an item
	* player.c: set immutable time to 60 instead of 50 when
	  starting/restarting
	* xsoldier.h: changed the argument of the srand() macro from s to x
	  because the word "srand" contains the letter "s"
	* boss.c: homing shot of the last boss is faster
	* boss.c, enemy.c, enemyshot.c: fixed window edge detection
	* boss.c: fixed homing shot of boss 2

main.c
/* xsoldier, a shoot 'em up game with "not shooting" bonus
 * Copyright (C) 1997 Yuusuke HASHIMOTO <s945750@educ.info.kanagawa-u.ac.jp>
 * Copyright (C) 2002 Oohara Yuuma  <oohara@libra.interq.or.jp>
 *
 * This is a copyleft program.  See the file LICENSE for details.
 */
/* $Id: main.c,v 1.31 2006/09/16 09:20:54 oohara Exp $ */

/* DEBUG and JSTK are defined in config.h */
#include <config.h>
/* time */
#include <time.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/keysym.h>

#include "xsoldier.h"
#include "manage.h"
#include "opening.h"
#include "game.h"
#include "ending.h"
#include "player.h"
/* DeleteAllStar */
#include "star.h"
#include "score.h"
#include "wait.h"
#include "graphic.h"
#include "input.h"

/* define all "extern" here */
#define EXTERN_DEF
#include "extern.h"

/* local functions */
static void arginit(int argc, char *argv[]);
static void usage(void);

static void init(void);
static void clean(void);
static void sig_handle(int arg);

/* file-global variables */
static char command[32];
static int scoreOK = True;
static int putscore = False;
static int w_time = -1;

static int maxlevel_temp = False;
static int window_width = 500;
static int start_stage = 1;
static int start_loop = 1;
static int start_ship = 5;
static int maxpower_temp = False;
static int nopausemessage_temp = False;

int main(int argc, char *argv[])
{
  display[0] = '\0';
  manage = NULL;

    arginit(argc,argv);

    player = NewPlayerData();
    if (putscore == True)
    {
	int i;
	printf("      --- top 10 soldiers ---\n");
	printf("Name                 Stage Score\n");
	for (i=1; i<=10; i++)
	    printf("%-20.20s %2d-%2d %8d\n",
		   player->Rec[i].name,
		   player->Rec[i].loop,player->Rec[i].stage,
		   player->Rec[i].score);

	exit(0);
    }
    if (scoreOK == False)
	fprintf(stderr,"caution: no ranking mode!\n");

    init();
    manage = NewManage(256,256);
    if (maxlevel_temp == True)
      manage->flag_maxlevel = True;
    if (maxpower_temp == True)
      manage->start_power = 30;
    else
      manage->start_power = 0;

    if (nopausemessage_temp == True)
      manage->flag_nopausemessage = True;
    
#ifndef HAVE_LIBSDL
    XMapWindow(dpy,win);
    XMapWindow(dpy,root);

    XFlush(dpy);
#endif /* not HAVE_LIBSDL */

    while (!(manage->program_should_quit))
    {
	keymask = 0;

	player->Rec[0].score = 0;
	player->Ships = start_ship;
	player->Next = FIRST1UP;
	manage->Loop = start_loop;
        manage->Stage = start_stage;
        
	if (Opening() == -1)
	    break;

	while (1)
	{
	    keymask = 0;
	    ResetManage(manage);

	    if (mainLoop() == 0)
              /* game over */
		break;
	    Ending();
	    manage->Loop++;
            manage->Stage = 1;
            if (manage->Loop >3)
              break;
	}

	if (scoreOK == True)
	{
	    if (MergeHiscore(player) > 0)
		WriteHiscore(player);
	}
    }

    DeleteManage(manage);
    free(player);
    clean();
    return 0;
}

static void arginit(int argc, char *argv[])
{
    char *argv0;
    int i;
    int a;
#ifdef DEBUG
    int b;
#endif /* DEBUG */

    if ((argv0=strrchr(argv[0],'/')) != NULL)
        argv0++;
    else
        argv0 = argv[0];
    /* copy command name */
    strncpy(command,argv0, sizeof(command)-1);

    for (i=1; i<argc; i++)
    {
	if ((strcmp(argv[i],"-display")==0) || (strcmp(argv[i],"-d")==0))
	{
	    if (i < argc-1)
            {
              strncpy(display,argv[i + 1],sizeof(display));
              if (display[sizeof(display) - 1] != '\0')
              {
                fprintf(stderr, "warning: display name (arg %d) is too long, ",
                        i + 1);
                display[sizeof(display) - 1] = '\0';
                fprintf(stderr, "truncated to %d chars\n",
                        sizeof(display) - 1);
              }
              i++;
            }
            else
            {
              fprintf(stderr, "no display specified for -display (arg %d)\n",
                      i);
              exit(1);
            }
	}
        else if ((strcmp(argv[i],"-wait")==0) || (strcmp(argv[i],"-w")==0))
	{
	    if (i < argc-1)
	    {
              if ((sscanf(argv[i+1], "%d", &a) == 1) && (a >= 1))
              {
                w_time = a;
                if (w_time > WAIT)
                  scoreOK = False;
                i++;
              }
              else
              {
                fprintf(stderr, "wait (arg %d) must be a positive integer.\n",
                        i + 1);
                exit(1);
              }
            }
            else
            {
              fprintf(stderr, "no wait specified for -wait (arg %d)\n", i);
              exit(1);
            }
	}
        else if ((strcmp(argv[i],"-cmap")==0) || (strcmp(argv[i],"-c")==0))
        {
          /* ignore */
          ; 
	}
        else if ((strcmp(argv[i],"-score")==0) || (strcmp(argv[i],"-s")==0))
	    putscore = True;
        else if ((strcmp(argv[i],"-help")==0) || (strcmp(argv[i],"-h")==0))
            usage();
        else if (strcmp(argv[i],"-maxlevel")==0)
	    maxlevel_temp = True;
        else if ((strcmp(argv[i],"-width")==0))
        {
            if (i < argc-1)
            {
                if ((sscanf(argv[i+1], "%d", &a) == 1) && (a >= 500))
                {
                    window_width = a;
                    i++;
                }
                else
                {
                    fprintf(stderr,
                            "width (arg %d) must be a positive integer and greater than 500\n",
                            i + 1);
                    exit(1);
                }
            }
            else
            {
                fprintf(stderr, "no width specified for -width (arg %d)\n", i);
                exit(1);
            }
        }
        else if (strcmp(argv[i],"-nopausemessage")==0)
	    nopausemessage_temp = True;
#ifdef DEBUG
        else if (strcmp(argv[i],"-stage") == 0)
        {
          if (i + 1 < argc)
          {
            if (sscanf(argv[i+1], "%d-%d", &a, &b) == 2)
            {
              if ((a < 1) || (a > 3))
              {
                fprintf(stderr, "the loop number (arg %d) must be between "
                        "1 and 3\n", i + 1);
                exit(1);
              }
              else if ((b < 1) || (b > 8))
              {
                fprintf(stderr, "the stage number (arg %d) must be between "
                        "1 and 8\n", i + 1);
                exit(1);
              }
              else
              {
                scoreOK = False;
                start_loop = a;
                start_stage = b;
                i++;
              }
            }
            else if (sscanf(argv[i+1], "%d", &b) == 1)
            {
              if ((b >= 1) && (b <= 8))
              {
                scoreOK = False;
                start_stage = b;
                i++;
              }
              else
              {
                fprintf(stderr, "the stage number (arg %d) must be between "
                        "1 and 8\n", i + 1);
                exit(1);
              }
            }
            else
            {
              fprintf(stderr, "strange stage number (arg %d) for -stage\n",
                      i + 1);
              exit(1);
            }
          }
          else
          {
            fprintf(stderr, "missing stage number for -stage (arg %d)\n", i);
            exit(1);
          }
        }
        else if (strcmp(argv[i],"-loop") == 0)
        {
          if (i + 1 < argc)
          {
            if (sscanf(argv[i+1], "%d", &a) == 1)
            {
              if ((a >= 1) && (a <= 3))
              {
                scoreOK = False;
                start_loop = a;
                i++;
              }
              else
              {
                fprintf(stderr, "loop number (arg %d) must be between "
                        "1 and 3\n", i + 1);
                exit(1);
              }
            }
            else
            {
              fprintf(stderr, "strange loop number (arg %d) for -loop\n",
                      i + 1);
              exit(1);
            }
          }
          else
          {
            fprintf(stderr, "missing loop number for -loop (arg %d)\n", i);
            exit(1);
          }
        }
        else if (strcmp(argv[i],"-ship") == 0)
        {
          if (i + 1 < argc)
          {
            if (sscanf(argv[i+1], "%d", &a) == 1)
            {
              if ((a >= 0) && (a <= 99))
              {
                scoreOK = False;
                start_ship = a;
                i++;
              }
              else
              {
                fprintf(stderr, "ship number (arg %d) must be between "
                        "0 and 99\n", i + 1);
                exit(1);
              }
            } 
            else
            {
              fprintf(stderr, "strange ship number (arg %d) for -ship\n",
                      i + 1);
              exit(1);
            }
          }
          else
          {
            fprintf(stderr, "missing ship number for -ship (arg %d)\n", i);
            exit(1);
          }
        }
        else if (strcmp(argv[i],"-maxpower")==0)
        {
          scoreOK = False;
          maxpower_temp = True;
        }
#endif /* DEBUG */
        else
        {
          fprintf(stderr, "strange option (arg %d)\n", i);
          exit(1);
        }
    }
}

static void usage(void)
{
    fprintf(stderr,"Usage: %s [-display <name>] [-wait <n>] [-score] [-help] [-maxlevel]",command);
#ifdef DEBUG
    fprintf(stderr, "[-stage <L-S>] [-stage <S>] [-loop <L>] [-ship <N>] [maxpower]");
#endif /* DEBUG*/
    fprintf(stderr, "\n");

    fprintf(stderr,"\t display <name>   ... name of the X server to play game.\n");
    fprintf(stderr,"\t wait <n>         ... set wait time. (no ranking) <default is %d>\n",WAIT);
    fprintf(stderr,"\t width <n>        ... set window width <default is %d>\n",window_width);
    fprintf(stderr,"\t score            ... print top 10 soldiers.\n");
    fprintf(stderr,"\t help             ... this message.\n");
    fprintf(stderr,"\t maxlevel         ... force max level enemy attack.\n");
#ifdef DEBUG
    fprintf(stderr,"debug-only options\n");
    fprintf(stderr,"\t stage <L-S>      ... start stage S [1-8] of loop L [1-3]. (no ranking)\n");
    fprintf(stderr,"\t stage <S>        ... start stage S [1-8]. (no ranking)\n");
    fprintf(stderr,"\t loop <L>         ... start loop L [1-3]. (no ranking)\n");
    fprintf(stderr,"\t ship <n>         ... start with n [0-99] ships. (no ranking)\n");
    fprintf(stderr,"\t maxpower         ... start with max power. (no ranking)\n");
#endif /* DEBUG */
    exit(0);
}

static void init(void)
{
    struct itimerval value, ovalue;
    struct sigaction sig_act;

    /* set wait */
    signal_delivered = 1;
    if (w_time < 0)
	waittime = WAIT;
    else
	waittime = w_time;

    memset(&sig_act, 0, sizeof(sig_act));
    sig_act.sa_handler = sig_handle;
    sigaction(SIGALRM, &sig_act, 0);
    value.it_interval.tv_sec = 0;
    value.it_interval.tv_usec = waittime;
    value.it_value.tv_sec = 0;
    value.it_value.tv_usec = waittime;
    setitimer(ITIMER_REAL, &value, &ovalue);

    DefaultFieldW = 500;
    FieldW  = window_width;
    FieldH  = 650;

    /* you must call graphic_init() first because it calls SDL_Init */
    graphic_init(FieldW);
    input_init();

    srand((unsigned)time(NULL));
}

static void clean(void)
{
    graphic_finish();
}

static void sig_handle(int arg)
{
    signal_delivered = 1;
}
